p2p/src/network/pubsub/p2p_network_pubsub_reducer.rs

```diff
@@ -440,6 +437,126 @@ impl P2pNetworkPubsubState {
+                    super::ValidationResult::Valid => {
+                        match message_content {
+                            GossipNetMessageV2::NewState(block) => {
+                                let best_tip = BlockWithHash::try_new(block.clone())?;
+                                dispatcher.push(P2pPeerAction::BestTipUpdate { peer_id, best_tip });
+                                return Ok(());
+                            }
+                            GossipNetMessageV2::TransactionPoolDiff { message, nonce } => {
+                                let nonce = nonce.as_u32();
+                                for transaction in message.0 {
+                                    dispatcher.push(P2pChannelsTransactionAction::Libp2pReceived {
+                                        peer_id,
+                                        transaction: Box::new(transaction),
+                                        nonce,
+                                    });
+                                }
+                            }
+                            GossipNetMessageV2::SnarkPoolDiff {
+                                message:
+                                    NetworkPoolSnarkPoolDiffVersionedStableV2::AddSolvedWork(work),
+                                nonce,
+                            } => {
+                                dispatcher.push(P2pChannelsSnarkAction::Libp2pReceived {
+                                    peer_id,
+                                    snark: Box::new(work.1.into()),
+                                    nonce: nonce.as_u32(),
+                                });
+                            }
+                            _ => {}
+                        }
+
+                        Self::broadcast(dispatcher, global_state)
+                    }
+                    super::ValidationResult::Reject => {
+                        // TODO: add error variants for transactions and snarks
+                        dispatcher.push(P2pDisconnectionAction::Init {
+                            peer_id,
+                            reason: P2pDisconnectionReason::BlockVerifyError,
+                        });
+                        Ok(())
+                    }
+                    super::ValidationResult::Ignore => Ok(()),
+                }
+            }
+            P2pNetworkPubsubAction::BroadcastAcceptedBlock { hash } => {
+                let Some(message) = pubsub_state.block_messages.remove(&hash) else {
+                    bug_condition!("Block message not found for: {}", hash);
+                    return Ok(());
+                };
+
+                let P2pNetworkPubsubBlockMessage {
+                    message_id,
+                    peer_id,
+                    ..
+                } = message;
+
+                let Some(message_id) = message_id else {
+                    return Ok(());
+                };
+
+                let Some(message) = pubsub_state.mcache.map.get(&message_id) else {
```

Comment:
on which cases is this (and the above `message_id`) expected to be `None` ? Should it be a bug condition? Skipping the broadcasting of a block only because we don't have in the cache doesn't look right. But if there is a situation on which this would be the expected behavior please add a comment explaining that.

Reply:
I see it can be `None` if `compute_message_id` fails to produce an id, but is that something that should fail silently? shouldn't it be an error instead?

The way it works right now it just fails silently many steps before, and then it silently fails to broadcast the message here, thats not very good.

Reply:
`compute_message_id` should return a `Result` instead. What I don't understand is under which conditions the peer id can be invalid. That doesn't depend on the input right? if that happens, it is because of a bug? (in such case it should be a `bug_condition`, otherwise, it should interrupt the flow and dispatch an error action when it happens.

Reply:
Other possibility is that it is impossible for it to be invalid. If that is the case I would revert `compute_message_id` to use `unwrap`/`expect` like before and add a comment explaining that it cannot fail

Reply:
Btw a case in which we are not going to have anything in the cache is when the best tip comes from webrtc.

Reply:
same for the `pubsub_state.block_messages.remove(&hash)` above I think

Reply:
(the webrtc->libp2p aspect is out of scope of this PR unless you think it is easy to solve, but if thats not the case I would add a TODO here)